Discard decoded image data in Web processes with no foreground tabs

heycam opened the pull request:
#### baeabe11daecb0ce0df709e88ae0ef821f7a2bcb
<pre>
Discard decoded image data in Web processes with no foreground tabs
<a href="https://bugs.webkit.org/show_bug.cgi?id=254768">https://bugs.webkit.org/show_bug.cgi?id=254768</a>
&lt;rdar://problem/107437769&gt;

Reviewed by Said Abou-Hallawa.

To reduce memory footprint, this patch discards decoded image data in
Web processes with no foreground tabs.

Destroying the decoded data means there is a risk of images not
rendering next time we paint, if we had decided to async decode (due to
the images being large, or referenced with &lt;img decode=async&gt;, etc.).
This can manifest in a couple of ways:

1. When switching back to a tab that&apos;s had its decoded image data
   destroyed, we won&apos;t see the effect immediately, since we&apos;ll show the
   page snapshot without painting the page again. By the next time we
   do need to paint (e.g. after the user interacts with the page in some
   way), the image may briefly disappear while we wait for the async
   decode.

2. When hovering over the tab strip, Safari generates thumbnails showing
   the contents of the tabs. Although the image painting is performed
   with PaintBehavior::Snapshotting, this does not currently override
   &lt;img decode=async&gt;, so any previously decoded &lt;img decode=async&gt;
   image will appear blank in the thumbnail.

To avoid these issues, we tell the RemoteLayerTreeDrawingArea to decode
images synchronously in the next rendering update, and adjust
RenderBoxModelObject::decodingModeForImageDraw to prioritize
PaintBehavior::Snapshotting over &lt;img decode=async&gt;.

RemoteLayerTreeDrawingArea can perform no-paint rendering updates when
switching back to a background tab, so we wait until we do some painting
(a backing store has a non-empty dirty region) before clearing
m_nextRenderingUpdateRequiresSynchronousImageDecoding.

WebProcess::m_nonVisibleProcessEarlyMemoryCleanupTimer is changed to a
DeferrableOneShotTimer. If a tab is in the background, and we have
already destroyed the image decoded data, we schedule the timer again.
This avoids the image data persisting if the page thumbnails are
generated. If a tab is in the background, and we haven&apos;t yet destroyed
the image decoded data (i.e., if the timer is pending), then we defer
the timer when an image is painted. This avoids continuously destroying
then regenerating the decoded data, if the page is repeatedly drawing
images in the background.

When the timer fires, we also tell the RemoteResourceCacheProxy to clear
all its image resources at once, rather than have each NativeImage in
its destructor tell the RemoteResourceCacheProxy to remove a single
image resource, to reduce IPC traffic.

For now, we only do this when UI side compositing is enabled, but it
would also be beneficial in other configurations.

* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::decodingModeForImageDraw const):
* Source/WebKit/GPUProcess/graphics/QualifiedResourceHeap.h:
(WebKit::QualifiedResourceHeap::releaseAllImageResources):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::releaseAllImageResources):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.cpp:
(WebKit::RemoteResourceCache::releaseAllImageResources):
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.mm:
(WebKit::RemoteLayerBackingStoreCollection::paintReachableBackingStoreContents):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::releaseAllImageResources):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::recordNativeImageUse):
(WebKit::RemoteResourceCacheProxy::releaseAllImageResources):
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.h:
* Source/WebKit/WebProcess/WebPage/DrawingArea.h:
(WebKit::DrawingArea::setNextRenderingUpdateRequiresSynchronousImageDecoding):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.h:
(WebKit::RemoteLayerTreeContext::setNextRenderingUpdateRequiresSynchronousImageDecoding):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.mm:
(WebKit::RemoteLayerTreeContext::buildTransaction):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::setNextRenderingUpdateRequiresSynchronousImageDecoding):
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
(WebKit::RemoteLayerTreeDrawingArea::activityStateDidChange):
(WebKit::RemoteLayerTreeDrawingArea::dispatchAfterEnsuringDrawing):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::willDestroyDecodedDataForAllImages):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::WebProcess):
(WebKit::WebProcess::pageDidEnterWindow):
(WebKit::WebProcess::pageWillLeaveWindow):
(WebKit::WebProcess::nonVisibleProcessEarlyMemoryCleanupTimerFired):
(WebKit::WebProcess::destroyDecodedDataForAllImages):
(WebKit::WebProcess::deferNonVisibleProcessEarlyMemoryCleanupTimer):
* Source/WebKit/WebProcess/WebProcess.h:

Canonical link: <a href="https://commits.webkit.org/262585@main">https://commits.webkit.org/262585@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7ece1af1b1a9d9af5aadedcdfb25051c9949cf11

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1956 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1987 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2049 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2883 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2044 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1932 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2064 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2033 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1804 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1975 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1763 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1755 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2729 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1754 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1737 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1650 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1800 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1784 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2901 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1812 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/1611 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1739 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1748 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/482 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1903 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/platform/graphics/BitmapImage.cpp:
```diff
@@ -250,8 +252,16 @@ ImageDrawResult BitmapImage::draw(GraphicsContext& context, const FloatRect& des
+        LOG(Images, "BitmapImage::%s - %p - url: %s [forcing synchronous decode]", __FUNCTION__, this, sourceURL().string().utf8().data());
+        decodingMode = DecodingMode::Synchronous;
+        setFrameRequiresSynchronousDecode(m_currentFrame, s_requireSynchronousRedecoding);
+    }
```
This is kind of a hack. We should not have set the `decodingMode` to `Asynchronous` from the beginning. Please move this logic to `RenderBoxModelObject::decodingModeForImageDraw()` and let it make the decision itself. `BitmapImage` and `ImageFrame` should not know about all of this.

heycam replied:
If (as I write in the previous comment) each image does need to track whether its next decode must be synchronous, should I iterate over the frame tree to set a flag on all the RenderBoxModelObjects? We really only need to do this for images which have already been decoded and have had their decoded data destroyed, but I'm not sure if there's a good way to know that from the RenderBoxModelObject.

heycam replied:
Oh, I guess we could inform the RenderBoxModelObject through the CachedResourceClient interface, if we're already iterating over all the CachedImages to discard the data.

shallawa commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -1677,12 +1681,29 @@ void WebProcess::nonVisibleProcessEarlyMemoryCleanupTimerFired()
 #endif
 }
 
+void WebProcess::destroyDecodedImageData()
```
I think the name of this function should reflect it destroys the decoded data for "all" images. Maybe `destroyDecodedDataForAllImages()` is fine to be reused.

shallawa commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -1677,12 +1681,29 @@ void WebProcess::nonVisibleProcessEarlyMemoryCleanupTimerFired()
+        page->releaseAllRemoteImageResources();
+
+    MemoryCache::singleton().destroyDecodedDataForAllImages();
+    BitmapImage::setRequireSynchronousRedecoding(true);
```
The meaning of this setting should be: Next time this page is brought to foreground, decode all images in viewport synchronously. In other words make `RenderBoxModelObject::decodingModeForImageDraw()` ignore this if-statement till the page enters the window.
```
    if (paintInfo.paintBehavior.contains(PaintBehavior::TileFirstPaint))
        return DecodingMode::Asynchronous;
```

heycam replied:
What about the case where `<img decoding=async>`? I think we also want to make that decode synchronously in this case, otherwise we will see a flicker of unpainted image when we go from the page snapshot (which does show the image) to the next actual paint (when we start the async decode).

shallawa replied:
Yes this is true and this what we do `RenderBoxModelObject::decodingModeForImageDraw()` if we passed the `PaintBehavior::TileFirstPaint` milestone and we are inside the viewport. It returns `DecodingMode::Synchronous`. So all what we need is to pretend the layer has been painted previously.

heycam replied:
This block, though, is above the check for `TileFirstPaint`:

```
    if (is<HTMLImageElement>(element())) {
        auto decodingMode = downcast<HTMLImageElement>(*element()).decodingMode();
        if (decodingMode == DecodingMode::Asynchronous)
            return DecodingMode::Asynchronous;
        if (decodingMode == DecodingMode::Synchronous)
            return DecodingMode::Synchronous;
    }
```

heycam replied:
So I think that means we'll decode async for `<img decode=async>` even when ignoring `TileFirstPaint`.

shallawa commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -1677,12 +1681,29 @@ void WebProcess::nonVisibleProcessEarlyMemoryCleanupTimerFired()
+void WebProcess::destroyDecodedImageData()
+{
+    for (auto& page : m_pageMap.values())
+        page->releaseAllRemoteImageResources();
```
It is difficult to keep track with adding and removing "remote" from the name of these functions.

1. This function calls WebPage::releaseAllRemoteImageResources()
2. Which calls RemoteResourceCacheProxy::releaseAllNativeImages()
3. Which calls RemoteRenderingBackendProxy::releaseAllRemoteNativeImages()
4. Which sends the message RemoteRenderingBackend::ReleaseAllNativeImages()

heycam replied:
I will make them all consistent.

shallawa commented on Source/WebCore/platform/graphics/ImageFrame.cpp:
```diff
@@ -85,6 +86,7 @@ unsigned ImageFrame::clearImage()
     m_nativeImage->clearSubimages();
     m_nativeImage = nullptr;
     m_decodingOptions = DecodingOptions();
+    m_requiresSynchronousDecode = true;
```
I do not this is needed. And I think it is incorrect as well. We clear the image frame under memory pressure. The `decodingMode` should be specified by the client of the image. We should not assume once the frame is deleted, it has to be synchronously decoded always. 

The logic to reset this flag once `BitmapImage::s_requireSynchronousRedecoding` is reset to false is not reflected in the name and it is not obvious.

heycam replied:
Without this flag on the ImageFrame, there's no easy way to track whether this is the first paint since the tab came back to the foreground. We set the static s_requireSynchronousRedecode back to false when the tab goes into the foreground, but the first repaint might be some time later, because we've shown the page snapshot initially. So by the time the user interacts with the page to cause an actual first repaint, we don't know that we need to do decode synchronously.

shallawa commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.mm:
```diff
@@ -63,6 +63,15 @@
+            return true;
+    }
+    return false;
+}
```
Can't we remove this function and make it the output of `RemoteLayerBackingStoreCollection::paintReachableBackingStoreContents()` since these two functions are called together. So `paintReachableBackingStoreContents()` will paint and will return true if it has already painted something.

shallawa commented on Source/WebKit/WebProcess/WebPage/DrawingArea.h:
```diff
@@ -165,6 +165,8 @@ class DrawingArea : public IPC::MessageReceiver, public WebCore::DisplayRefreshM
     virtual void adoptLayersFromDrawingArea(DrawingArea&) { }
     virtual void adoptDisplayRefreshMonitorsFromDrawingArea(DrawingArea&) { }
 
+    virtual void setNextRenderingUpdateRequiresSynchronousImageDecoding() { }
```
If this function is going to be accessed from the base class, I'd rather have the flag also be in the base class and do not make this function virtual. For me virtual function means completely different implementation/meaning in the superclass. But  in this case the superclass is doing the obvious thing.

shallawa commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.mm:
```diff
@@ -188,6 +188,7 @@
     m_currentTransaction = nullptr;
 
+    transaction.setPaintedAnyBackingStore(m_backingStoreCollection->hasAnyReachableBackingStoreWithNonEmptyDirtyRegion());
     m_backingStoreCollection->paintReachableBackingStoreContents();
```
```
bool paintedAnyBackingStore = m_backingStoreCollection->paintReachableBackingStoreContents();
transaction.setPaintedAnyBackingStore(paintedAnyBackingStore);
```

shallawa commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -1677,12 +1677,31 @@ void WebProcess::nonVisibleProcessEarlyMemoryCleanupTimerFired()
+    for (auto& page : m_pageMap.values()) {
+        page->releaseAllRemoteImageResources();
+        if (auto* drawingArea = page->drawingArea())
+            drawingArea->setNextRenderingUpdateRequiresSynchronousImageDecoding();
```
Can this be part of WebPage::releaseAllRemoteImageResources()? The WebPage is destroying the decoded data so it makes sense to mark its DrawingArea for synchronous image decoding.

heycam replied:
It's really the destroyDecodedDataForAllImages() call in WebProcess::destroyDecodedDataForAllImages that is the thing that needs this flag set, not the remote resource releasing (which would be fine even without setting the flag). So I think it makes more sense here.

heycam replied:
I guess it could go either way; the destroyDecodedDataForAllImages works on all pages' image resources (since it's the process global MemoryCache it's working on).

shallawa commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
```diff
@@ -401,6 +404,7 @@ class RemoteLayerTreeTransaction {
     bool m_viewportMetaTagWidthWasExplicit { false };
     bool m_viewportMetaTagCameFromImageDocument { false };
     bool m_isInStableState { false };
+    bool m_paintedAnyBackingStore { false }; // Set by buildTransaction but not IPCed.
```
Is transaction the right place to hold this flag since it is not going to be encoded? I think WebPage might be a better place to have this flag. It is set by RemoteLayerTreeContext and consumed by RemoteLayerTreeDrawingArea and both has access to WebPage.

heycam replied:
I can move it there, but I feel it's a little strange, particularly because the value doesn't need to be used outside RemoteLayerTreeDrawingArea::updateRendering. Maybe it should be a return value or an outparam from buildTransaction?

shallawa commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.mm:
```diff
@@ -188,7 +188,9 @@
     m_currentTransaction = nullptr;
 
-    m_backingStoreCollection->paintReachableBackingStoreContents();
+    bool paintedAnyBackingStore = m_backingStoreCollection->paintReachableBackingStoreContents();
+    if (paintedAnyBackingStore)
+        m_nextRenderingUpdateRequiresSynchronousImageDecoding = false;
```
I think the local variable is not really needed. 
```
if (m_backingStoreCollection->paintReachableBackingStoreContents())
    m_nextRenderingUpdateRequiresSynchronousImageDecoding = false;
```

shallawa commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -1703,12 +1703,35 @@ void WebProcess::nonVisibleProcessEarlyMemoryCleanupTimerFired()
+void WebProcess::destroyDecodedDataForAllImages()
+{
+#if PLATFORM(COCOA)
+    for (auto& page : m_pageMap.values()) {
+        if (!page->isUsingUISideCompositing())
+            return;
+    }
```
I do not understand this code. `DrawingArea::setNextRenderingUpdateRequiresSynchronousImageDecoding()` is virtual and it is only implemented in RemoteLayerTreeDrawingArea. So the for-loop below is doing nothing unless the page has a `DrawingArea` of type `RemoteLayerTreeDrawingArea`. Also the mix of having `PLATFORM(COCOA)` and  `isUsingUISideCompositing()` in this function and having `ENABLE(GPU_PROCESS)` in `WebPage::willDestroyDecodedDataForAllImages()` makes it difficult to see when this code can be enabled.

heycam replied:
I think we want to destroy the decoded image data only when UI side compositing is being used. Otherwise, because `setNextRenderingUpdateRequiresSynchronousImageDecoding` does nothing on a `TiledCoreAnimationDrawingArea`, we will see a flash of unrendered images. Once we implement `setNextRenderingUpdateRequiresSynchronousImageDecoding` on `TiledCoreAnimationDrawingArea` we can remove this check.

An alternative would be to make `willDestroyDecodedDataForAllImages` return whether the decoded data destruction is allowed.

The `PLATFORM(COCOA)` is indeed not needed. (In a previous iteration of the patch I didn't upload, I was checking the drawing area type, and `DrawingAreaType::RemoteLayerTree` is only defined on `PLATFORM(COCOA)`.)

heycam committed:
Discard decoded image data in Web processes with no foreground tabs
https://bugs.webkit.org/show_bug.cgi?id=254768
<rdar://problem/107437769>

Reviewed by Said Abou-Hallawa.

To reduce memory footprint, this patch discards decoded image data in
Web processes with no foreground tabs.

Destroying the decoded data means there is a risk of images not
rendering next time we paint, if we had decided to async decode (due to
the images being large, or referenced with <img decode=async>, etc.).
This can manifest in a couple of ways:

1. When switching back to a tab that's had its decoded image data
   destroyed, we won't see the effect immediately, since we'll show the
   page snapshot without painting the page again. By the next time we
   do need to paint (e.g. after the user interacts with the page in some
   way), the image may briefly disappear while we wait for the async
   decode.

2. When hovering over the tab strip, Safari generates thumbnails showing
   the contents of the tabs. Although the image painting is performed
   with PaintBehavior::Snapshotting, this does not currently override
   <img decode=async>, so any previously decoded <img decode=async>
   image will appear blank in the thumbnail.

To avoid these issues, we tell the RemoteLayerTreeDrawingArea to decode
images synchronously in the next rendering update, and adjust
RenderBoxModelObject::decodingModeForImageDraw to prioritize
PaintBehavior::Snapshotting over <img decode=async>.

RemoteLayerTreeDrawingArea can perform no-paint rendering updates when
switching back to a background tab, so we wait until we do some painting
(a backing store has a non-empty dirty region) before clearing
m_nextRenderingUpdateRequiresSynchronousImageDecoding.

WebProcess::m_nonVisibleProcessEarlyMemoryCleanupTimer is changed to a
DeferrableOneShotTimer. If a tab is in the background, and we have
already destroyed the image decoded data, we schedule the timer again.
This avoids the image data persisting if the page thumbnails are
generated. If a tab is in the background, and we haven't yet destroyed
the image decoded data (i.e., if the timer is pending), then we defer
the timer when an image is painted. This avoids continuously destroying
then regenerating the decoded data, if the page is repeatedly drawing
images in the background.

When the timer fires, we also tell the RemoteResourceCacheProxy to clear
all its image resources at once, rather than have each NativeImage in
its destructor tell the RemoteResourceCacheProxy to remove a single
image resource, to reduce IPC traffic.

For now, we only do this when UI side compositing is enabled, but it
would also be beneficial in other configurations.

* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::decodingModeForImageDraw const):
* Source/WebKit/GPUProcess/graphics/QualifiedResourceHeap.h:
(WebKit::QualifiedResourceHeap::releaseAllImageResources):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::releaseAllImageResources):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.cpp:
(WebKit::RemoteResourceCache::releaseAllImageResources):
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.mm:
(WebKit::RemoteLayerBackingStoreCollection::paintReachableBackingStoreContents):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::releaseAllImageResources):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::recordNativeImageUse):
(WebKit::RemoteResourceCacheProxy::releaseAllImageResources):
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.h:
* Source/WebKit/WebProcess/WebPage/DrawingArea.h:
(WebKit::DrawingArea::setNextRenderingUpdateRequiresSynchronousImageDecoding):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.h:
(WebKit::RemoteLayerTreeContext::setNextRenderingUpdateRequiresSynchronousImageDecoding):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.mm:
(WebKit::RemoteLayerTreeContext::buildTransaction):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::setNextRenderingUpdateRequiresSynchronousImageDecoding):
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
(WebKit::RemoteLayerTreeDrawingArea::activityStateDidChange):
(WebKit::RemoteLayerTreeDrawingArea::dispatchAfterEnsuringDrawing):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::willDestroyDecodedDataForAllImages):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::WebProcess):
(WebKit::WebProcess::pageDidEnterWindow):
(WebKit::WebProcess::pageWillLeaveWindow):
(WebKit::WebProcess::nonVisibleProcessEarlyMemoryCleanupTimerFired):
(WebKit::WebProcess::destroyDecodedDataForAllImages):
(WebKit::WebProcess::deferNonVisibleProcessEarlyMemoryCleanupTimer):
* Source/WebKit/WebProcess/WebProcess.h:

Canonical link: https://commits.webkit.org/262585@main
